tests/test_threadpool_limits.py

```diff
@@ -0,0 +1,176 @@
+import re
+import ctypes
+import pytest
+
+
+from loky import cpu_count
+from loky.backend import threadpool_limits
+from loky.backend._thread_pool_limiters import get_threadpool_limits
+from loky.backend._thread_pool_limiters import _set_threadpool_limits
+from loky.backend._thread_pool_limiters import ALL_PREFIXES, ALL_USER_APIS
+
+from .utils import with_check_openmp_n_threads, libopenblas_paths
+
+
+def should_skip_module(module):
+    # Possible bug in getting maximum number of threads with OpenBLAS < 0.2.16
+    # and OpenBLAS does not expose its version before 0.3.4.
+    return module['internal_api'] == "openblas" and module['version'] is None
+
+
+@pytest.mark.parametrize("prefix", ALL_PREFIXES)
+def test_threadpool_limits(openblas_present, mkl_present, prefix):
+    old_limits = get_threadpool_limits()
+
+    prefix_found = len([1 for module in old_limits
+                        if prefix == module['prefix']])
+    old_limits = {clib['prefix']: clib['n_thread'] for clib in old_limits}
+
+    if not prefix_found:
+        have_mkl = len({'mkl_rt', 'libmkl_rt'}.intersection(old_limits)) > 0
+        if "mkl_rt" in prefix and mkl_present and not have_mkl:
+            raise RuntimeError("Could not load the MKL prefix")
+        elif prefix == "libopenblas" and openblas_present:
+            raise RuntimeError("Could not load the OpenBLAS prefix")
+        else:
+            pytest.skip("Need {} support".format(prefix))
+
+    new_limits = _set_threadpool_limits(limits={prefix: 1})
+    new_limits = {clib['prefix']: clib['n_thread'] for clib in new_limits}
+    assert new_limits[prefix] == 1
+
+    threadpool_limits(limits={prefix: 3})
+    new_limits = get_threadpool_limits()
+    new_limits = {clib['prefix']: clib['n_thread'] for clib in new_limits}
+    assert new_limits[prefix] in (3, cpu_count(), cpu_count() // 2)
+
+    threadpool_limits(limits=old_limits)
+    new_limits = get_threadpool_limits()
+    new_limits = {clib['prefix']: clib['n_thread'] for clib in new_limits}
+    assert new_limits[prefix] == old_limits[prefix]
+
+
+@pytest.mark.parametrize("user_api", (None, "blas", "openmp"))
+def test_set_threadpool_limits_apis(user_api):
+    # Check that the number of threads used by the multithreaded C-libs can be
+    # modified dynamically.
+
+    if user_api is None:
+        api_modules = ('blas', 'openmp')
+    else:
+        api_modules = (user_api,)
+
+    old_limits = get_threadpool_limits()
```

Comment:
wrap the rest of this test in a try / finally block to try to avoid side effects in case of an assertion failure in this specific test.